src/twisted/internet/test/test_endpoints.py

```diff
@@ -1687,6 +1688,412 @@ class WithResolver(proxyForInterface(
+        self.fakeResolverCalls.append((hostName, portNumber))
+        return self.fakeResolverReturns
+
+
+    def test_interface(self):
+        """
+        A L{endpoints._SimpleHostnameResolver} instance provides
+        L{interfaces.IHostnameResolver}.
+        """
+        self.assertTrue(verifyObject(interfaces.IHostnameResolver,
+                                     self.resolver))
+
+
+    def test_resolveNameFailure(self):
+        """
+        A resolution failure is logged with the name that failed to
+        resolve and the callable that tried to resolve it.  The
+        resolution receiver begins, receives no addresses, and
+        completes.
+        """
+        logs = []
+
+        @provider(ILogObserver)
+        def captureLogs(event):
+            logs.append(event)
+
+        globalLogPublisher.addObserver(captureLogs)
+        self.addCleanup(lambda: globalLogPublisher.removeObserver(captureLogs))
+
+        receiver = self.resolver.resolveHostName(self.receiver, "example.com")
+
+        self.assertIs(receiver, self.receiver)
+
+        self.fakeResolverReturns.errback(Exception())
+
+        self.assertEqual(1, len(logs))
+        self.assertEqual(1, len(self.flushLoggedErrors(Exception)))
+
+        [event] = logs
+        self.assertTrue(event.get("isError"))
+        self.assertTrue(event.get("name", "example.com"))
+        self.assertTrue(event.get("callable", repr(self.fakeResolver)))
+
+        self.assertEqual(1, len(self.resolutionBeganCalls))
+        self.assertEqual(self.resolutionBeganCalls[0].name, "example.com")
+        self.assertFalse(self.addressResolvedCalls)
+        self.assertEqual(1, self.resolutionCompleteCallCount)
+
+
+    def test_resolveNameDelivers(self):
+        """
+        The resolution receiver begins, and resolved hostnames are
+        delivered before it completes.
+        """
+        port = 80
+        ipv4Host = '1.2.3.4'
+        ipv6Host = '1::2::3::4'
+
+        receiver = self.resolver.resolveHostName(self.receiver, "example.com")
+
+        self.assertIs(receiver, self.receiver)
+
+        self.fakeResolverReturns.callback([
+            (AF_INET, SOCK_STREAM, IPPROTO_TCP, '', (ipv4Host, port)),
+            (AF_INET6, SOCK_STREAM, IPPROTO_TCP, '', (ipv6Host, port)),
+        ])
+
+        self.assertEqual(1, len(self.resolutionBeganCalls))
+        self.assertEqual(self.resolutionBeganCalls[0].name, "example.com")
+        self.assertEqual(self.addressResolvedCalls, [
+            IPv4Address("TCP", ipv4Host, port),
+            IPv6Address("TCP", ipv6Host, port)
+        ])
+        self.assertEqual(self.resolutionCompleteCallCount, 1)
+
+
+
+class HostnameEndpointFallbackNameResolutionTests(unittest.TestCase):
+    """
+    Verify that L{HostnameEndpoint._fallbackNameResolution} defers a
+    name resolution call to a thread.
+    """
+
+    def test_fallbackNameResolution(self):
+        """
+        A L{Deferred} is returned that fires with the resoution of the
```

Comment:
Active voice not passive voice, "`_fallbackNameResolution` returns a `Deferred` that..."